### PR TITLE
header cleanup.

### PR DESCRIPTION
**Pull request by vondele**
remove unnecessary includes.

No functional change.

### CODE REVIEW

**Comment by Rocky640** [comment]
I checked with the tuning branch, 
https://github.com/glinscott/fishtest/wiki/Creating-my-first-test#tuning-with-spsa

and it works just fine.

Nice cleanup !

**Comment by snicolet** [comment]
At least some of these removals have been discussed before, and refused by the maintainers (Marco, if I remember correctly). The policy for SF code is that we don't want to rely on recursive includes, but keep a header as soon as at least one of the names it declares is used in the cpp file.

For instance, search.cpp uses an object of type MovePicker, so we keep `#include "movepick.h"`in top of search.cpp

**Comment by lucasart** [comment]
@snicolet: The master branch already uses recursive inclusion in many places. And not doing so would mean adding a lot of includes. So that was a bogus argument from Marco...

**Comment by mcostalba** [comment]
@vondele @lucasart the common c++ best practice seems to be:

```
In headers, always include anything that is needed for that header 
to be compiled with a clean .c/.cpp file.

In implementation files, always include everything that is directly used.
```

For a rationale see:

http://softwareengineering.stackexchange.com/questions/262019/is-it-good-practice-to-rely-on-headers-being-included-transitively

http://umich.edu/~eecs381/handouts/CppHeaderFileGuidelines.pdf  (rules #9, #11, #12)

http://www.acodersjourney.com/2016/05/top-10-c-header-file-mistakes-and-how-to-fix-them/  (rules #4, 

https://google.github.io/styleguide/cppguide.html#Names_and_Order_of_Includes  (Google C++ style guide)

And there are also more....

So if, as Lucas says, master branch already uses recursive inclusion in many places, I would gladly commit a patch that adds a lot if includes :-)

**Comment by erbsenzaehler** [comment]
I played with this tool https://github.com/include-what-you-use/include-what-you-use/blob/master/docs/WhyIWYU.md

Here are its recommendations. Some seem dubious, even to my eyes
[iwyu-stockfish.txt](https://github.com/official-stockfish/Stockfish/files/705853/iwyu-stockfish.txt)

**Comment by mcostalba** [comment]
@erbsenzaehler yes, for the most part is garbage, like

`#include <iosfwd>       // for string`

Theese tools could give a hint, but output should be **carfully** hand checked and filtered.

**Comment by vondele** [comment]
@mcostalba and @erbsenzaehler, I actually used iwyu for the analysis and decided to go for hand pruning instead. I don't think it is garbage for most part, it is enforcing the coding style google is using for their source base. <iosfwd> is for example useful for reducing compile time, IIUC (forward decl. instead of full header includes). On my ubuntu installation iwyu didn't deal with libstdc++ but if the tool works well, I think it would be a good style guide. (not that you can automatically cleanup the sources with ./fix_includes --comment < output).

**Comment by mcostalba** [comment]
@vondele IMO or you want to have consistent and clear usage of headers ot you want to be "smart" and think to reduce compile times.

If I use a string then, well, I want:

`#include <string>`

No more no less.

Moreover a lot standard libraries (stddef.h, stdint.h, etc) are not needed because most of them should be already included in #include "types.h".

If I include types.h in a cpp file or it is already included by the cpp  natural header, I don't want to repeat all the standard stuff: these headers will not change because are external to SF and fixed.

So to summarize:

- Standard includes already included in types.h do not need to be explicited again in ccp files
- If  types.h is already included in foo.h then I don't need to re-include it again in foo.cpp
- A generalization of the above: if file.h is already included in foo.h then don't include again in foo.cpp 

These simple and sound rules allow to greatly reduce the number of included in foo.cpp to the ones really needed and that we want to see explicitly.

**Comment by vondele** [comment]
@mcostalba the latest push removes files according to your three bullet points (I think)... 

It doesn't specify what to include explicitly (for example, should something that includes position.h include again bitboard.h) ? As a consequence I didn't add new headers.

**Comment by erbsenzaehler** [comment]
Ftw. I did run iwyu and polished the most obscure cases e.g. #include "syzygy/../search.h". I know that it is against the policy Marco suggested above and that it adds the standard includes over and over again...

https://github.com/erbsenzaehler/Stockfish/commit/f031adc2b6a6481e5e63e3c0f74a7a738b5c7d58

**Comment by mcostalba** [comment]
@vondele possibly there is a misunderstanding here.

My bullet points did not suggest to add not needed included in a header just to remove them from the cpp files.

You added in types.h
```
+#include <cstring>   // For std::memset
+#include <algorithm> // For std::max
+#include <vector>
+#include <string>
```

But this is **not** needed in types.h!!

Just to be clear:

- Any header should be self-sufficient, but should not include additional stuff apart from what is needed

See: http://stackoverflow.com/questions/1892043/self-sufficient-header-files-in-c-c

So adding external included in types.h and removing from the cpp files **does not help** in clarifying which headers are needed in a cpp file. Please remember our **target**:

- Include in each file the needed headers and no more than that
- Avoid adding useless or redundant headers to limit noise level and to have a better readability of what is really needed and **where** is needed.

So to summarize my point. The patch idea is good, but please update it so to avoid changing types.h.

**Comment by vondele** [comment]
@erbsenzaehler yes, I like that still of includes, because it is consistent and automatic. It might be useful to know explicitly that search.cpp depends explicitly on <atomic> <map> etc.

**Comment by vondele** [comment]
@mcostalba I moved those there to enable #972, which I think is a more concise and clean way to use a part of the standard types (and functions). types.h seems like the nature place to define that std::vector can be used throughout the code base as 'vector' (that is it is basic type in the code base). Feel free to close both PRs if you don't like that.

**Comment by mcostalba** [comment]
@vondele I am sorry, I appreciate your work but I still prefer to stick to include in the header what is needed there. I understand it is mostly a subjective decision and there is not a clear better way, but because of this there is not a clear reason to change and so in this cases we prefer to keep the current setup  because the new code does not bring a clear and objective plus: in case of even score current champion retains the title :-)
